text/2-aurgy.md

```diff
@@ -0,0 +1,135 @@
+* Generating a playlist
+* Playing music from the playlist
+
+Let's delve into these in more detail.
+
+#### Creating and managing a group/lobby
+
+In this project, users should be able to join a private, protected group. We could probably
+accomplish this using a standard lobby system: one person acts as the host and creates the
+lobby and receives a unique code; users who then enter this code can enter that particular
+lobby.
+
+We could probably set something like this up with a simple document database. Whenever a host
+creates a lobby, a new document is created with a unique identifying code. Then, if another
+user enters that code, their ID is added to the document. The document might look something
+like this:
+```json
+"<unique id>": {
+  "users": [
+    "<id>", "<id>", ...
+  ]
+}
+```
+
+In the interim, something like Firebase Firestore should be enough. It might be a good idea
+to switch to something else (like MongoDB perhaps) if we want to worry about scaling and
+whatnot.
+
+We could achieve persistent groups by just keeping these unique identifiers persistent, i.e.
+not deleting them. This way, the host and any other users can re-enter groups they are a part
+of by just entering their unique code.
+
+#### Generating a playlist
+
+Here's the meat of the project. We would make heavy use of the [Spotify API] and all of the
+information available about individual tracks here.
+
+First, we need a list of themes that the user can choose from. Some suggestions are "road trip",
+"party", or "kickback". Then, we would have to correlate each theme with a set of track features,
+such as genre.
+
+We would also need to understand the users' music preferences. Again, we would pull things like
+favorite (recent?) genres and figure out areas of strongest intersection between users.
+
+Finally, we want to take the features prescribed by the desired theme and the features that are
+most common among users, and pull songs that match as many of these features as possible.
+
+Once the playlist is generated, we can add it to the host's library.
+
+Playlist generation would of course be performed on the server side.
+
+[Spotify API]: https://developer.spotify.com/documentation/web-api/
+
+#### Playing music from the playlist
+
+Spotify offers a [playback SDK] to play music from Spotify on the client side. Combined with the
+[web API] for playback controls, we have a way for music to be played directly from our application.
+
+[playback SDK]: https://developer.spotify.com/documentation/web-playback-sdk/
+[web API]: https://developer.spotify.com/documentation/web-api/
+
+### What are the requirements for this project?
+
+The PWA itself will probably be built in React (possibly Next.js? I don't think it's necessary
+for this use case but @BryanPan342 what do you think?), and the backend will be Node.js-based.
+If we proceed with using Firestore for handling user lobbies to start with, then anyone with
+Firebase experience would be handy.
```

Comment:
We'll also need some custom Spotify auth bootstrapped onto Firebase since we're going to be storing Spotify Access/Refresh Tokens for users right

Reply:
Well potentially we could also just have the website prompt the user to log in again after an hour or however long it takes for Spotify auth to time out. I think that might be better for the MVP, but storing their access tokens is probably better in the long run. Unless i'm misunderstanding what you mean?

Reply:
Storing the refresh tokens like when they first spotify auth with us basically just allows us unlimited access to their spotify account so that they won't need to relogin to Spotify for our app. (also means that they won't have to like send their tokens to us with every request if it's more readily accessible in the backend)

Reply:
methinks we shouldn't store the access/refresh token on firebase? if anything they should be stored locally? More of a security/privacy concern than a dev one.

Reply:
If anything, not sure why we would need a Spotify auth with firebase or how that would look like? 

We can just have the user auth with Spotify, and have that trigger an update to firestore?

Reply:
Yeah I also don't know about storing their refresh tokens, from a security pov. Gonna hold off on changing anything re this in the RFC